### PR TITLE
Filter GitHub issues by label

### PR DESCRIPTION
**Pull request by dagda1**
## Hey, I just made a Pull Request!

Following on from the discussion in #13593 I have added optional `filterBy` and `orderBy` props to the components in the `github-issues plugin`.

The `filterBy` prop has the same optional fields as [IssueFilters object](https://docs.github.com/en/graphql/reference/input-objects#issuefilters), which is defined in the GitHub graphql schema.

The `orderBy` prop similarly has the same fields as the [IssueOrder object](https://docs.github.com/en/graphql/reference/input-objects#issueorder) defined in the Graphql schema.

The props can be assigned like this:

```ts
<GitHubIssuesCard
  filterBy={{
    labels: ['bug', 'enhancement'],
    states: ['OPEN']
  }}
  orderBy={{
    field: 'COMMENTS',
    direction: 'ASC',
  }}
/>
```

The props are optional, and the component can still be used as before:

```ts
<GitHubIssuesCard />
```

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/young-olives-cheer.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-github-issues': minor
```
Comment (jhaals) [inline .changeset/young-olives-cheer.md:2]:
```suggestion
'@backstage/plugin-github-issues': patch
```

Packages below version 0.x can add new features as patch :)

File: plugins/github-issues/src/api/gitHubIssuesApi.ts [hunk 2]
```diff
@@ -151,20 +189,50 @@ export const gitHubIssuesApi = (
...
+function formatFilterValue(value: IssuesFilters[keyof IssuesFilters]): string {
+  if (Array.isArray(value)) {
+    return `[ ${value.map(formatFilterValue).join(', ')}]`;
+  }
+
+  return typeof value === 'string' ? `\"${value}\"` : `${value}`;
+}
+
+function createFilterByClause(filterBy?: IssuesFilters): string {
+  if (typeof filterBy === 'undefined') {
```
Comment (jhaals) [inline plugins/github-issues/src/api/gitHubIssuesApi.ts:201]: If you want to check for it being unset or empty you can do.
```suggestion
  if (!filterBy) {
```

File: plugins/github-issues/src/api/gitHubIssuesApi.ts [hunk 3]
```diff
@@ -151,20 +189,50 @@ export const gitHubIssuesApi = (
...
+    return '';
+  }
+
+  return Object.entries(filterBy)
+    .flatMap(([field, value]) => {
+      if (typeof value === 'undefined') {
```
Comment (jhaals) [inline plugins/github-issues/src/api/gitHubIssuesApi.ts:207]:
```suggestion
      if (value) {
```

**Review by jhaals — CHANGES_REQUESTED** [review]
Thanks for the PR!

I left some comments. Hoping that the author @mrwolny also can take a look at this PR :)

File: plugins/github-issues/src/api/gitHubIssuesApi.ts (hunk 3 above)
Comment (dagda1) [reply plugins/github-issues/src/api/gitHubIssuesApi.ts:207]: @jhaals should that not be `if (!value) {` , i.e. not truthy?

Comment (jhaals) [reply plugins/github-issues/src/api/gitHubIssuesApi.ts:207]: yes you're right :)

File: plugins/github-issues/api-report.md [hunk 4]
```diff
@@ -27,7 +27,41 @@ export const gitHubIssuesPlugin: BackstagePlugin<
 export type GitHubIssuesProps = {
   itemsPerPage?: number;
   itemsPerRepo?: number;
+  filterBy?: IssuesFilters;
```
Comment (jhaals) [inline plugins/github-issues/api-report.md:30]: Let's prefix these new types with `Github` to make exports a bit more contained to this plugin

That makes it `GithubIssuesFilters` and `GithubIssuesOrdering`

File: plugins/github-issues/src/api/gitHubIssuesApi.ts [hunk 5]
```diff
@@ -151,20 +189,50 @@ export const gitHubIssuesApi = (
...
 
+function formatFilterValue(value: IssuesFilters[keyof IssuesFilters]): string {
+  if (Array.isArray(value)) {
+    return `[ ${value.map(formatFilterValue).join(', ')}]`;
+  }
+
+  return typeof value === 'string' ? `\"${value}\"` : `${value}`;
+}
+
+function createFilterByClause(filterBy?: IssuesFilters): string {
```
Comment (jhaals) [inline plugins/github-issues/src/api/gitHubIssuesApi.ts:200]: consider this optional as there are some tests but I would recommend adding some unit tests for this specific function to ensure it works as intended given that there are quite a few if statements inside it and that there's some ```typeof value === 'string' ? `\"${value}\"` : `${value}``` which I don't know if it's captured by the existing tests.

You can add test by exporting it with `export function createFilterByClause` but you don't have to add an export in the index.ts file.

**Review by jhaals — CHANGES_REQUESTED** [review]
Nice work, I found a few other tweaks :)

File: plugins/github-issues/api-report.md [hunk 6]
```diff
@@ -27,7 +51,17 @@ export const gitHubIssuesPlugin: BackstagePlugin<
 export type GitHubIssuesProps = {
   itemsPerPage?: number;
   itemsPerRepo?: number;
+  filterBy?: GithubIssuesFilters;
+  orderBy?: GithubIssuesOrdering;
 };
 
+// @public (undocumented)
+export interface IssuesByRepoOptions {
```
Comment (jhaals) [inline plugins/github-issues/api-report.md:59]:
```suggestion
export interface GithubIssuesByRepoOptions {
```

**Review by jhaals — CHANGES_REQUESTED** [review]
Nice work!

One more rename and then I think this is good to go 👏

**Review by jhaals — APPROVED** [review]
✨

**Comment by suuus** [comment]
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻

**Pull request merged**